crates/punktf-lib/src/deploy/executor.rs

```diff
@@ -665,94 +676,152 @@ where
-					);
+						return Ok(());
+					}
+				};
 
-					exec_dotfile.add_to_builder(
-						builder,
-						DotfileStatus::failed(format!("Failed to parse template: {}", err)),
-					);
+				let source = Source::file(exec_dotfile.source_path(), &content);
 
-					return Ok(());
-				}
-			};
+				let template = match Template::parse(source)
+					.with_context(|| format!("File: {}", exec_dotfile.source_path().display()))
+				{
+					Ok(template) => template,
+					Err(err) => {
+						log::error!(
+							"{}: Failed to parse template",
+							exec_dotfile.path().display()
+						);
 
-			let content = match template
-				.resolve(Some(profile.variables()), exec_dotfile.variables())
-				.with_context(|| format!("File: {}", exec_dotfile.source_path().display()))
-			{
-				Ok(template) => template,
-				Err(err) => {
-					log::error!(
-						"{}: Failed to resolve template",
-						exec_dotfile.path().display()
-					);
+						exec_dotfile.add_to_builder(
+							builder,
+							DotfileStatus::failed(format!("Failed to parse template: {}", err)),
+						);
 
-					exec_dotfile.add_to_builder(
-						builder,
-						DotfileStatus::failed(format!("Failed to resolve template: {}", err)),
-					);
+						return Ok(());
+					}
+				};
 
-					return Ok(());
+				let content = match template
+					.resolve(Some(profile.variables()), exec_dotfile.variables())
+					.with_context(|| format!("File: {}", exec_dotfile.source_path().display()))
+				{
+					Ok(template) => template,
+					Err(err) => {
+						log::error!(
+							"{}: Failed to resolve template",
+							exec_dotfile.path().display()
+						);
+
+						exec_dotfile.add_to_builder(
+							builder,
+							DotfileStatus::failed(format!("Failed to resolve template: {}", err)),
+						);
+
+						return Ok(());
+					}
+				};
+
+				log::trace!("{}: Resolved:\n{}", exec_dotfile.path().display(), content);
+
+				content
+			} else {
+				// We have some transformers to run on the content and thus can not straight copy the
+				// contents.
+				match std::fs::read_to_string(&exec_dotfile.source_path()) {
+					Ok(content) => content,
+					Err(err) => {
+						log::info!("{}: Failed to copy dotfile", exec_dotfile.path().display());
+
+						exec_dotfile.add_to_builder(
+							builder,
+							DotfileStatus::failed(format!("Failed to copy: {}", err)),
+						);
+
+						return Ok(());
+					}
 				}
 			};
 
-			log::trace!("{}: Resolved:\n{}", exec_dotfile.path().display(), content);
+			// Copy so we exec_dotfile is not referenced by this in case an error occurs.
+			let exec_transformers: Vec<_> = exec_dotfile.transformers().iter().copied().collect();
 
-			if !self.options.dry_run {
-				if let Err(err) = std::fs::write(&exec_dotfile.deploy_path(), content.as_bytes()) {
-					log::info!("{}: Failed to write content", exec_dotfile.path().display());
+			// Apply transformers.
+			// Order:
+			//   - Profile transformers
+			//   - then dotfile transformers
```

Comment:
i think the difference between "dotfile" and "profile" transformers is not clear. everytring is a dotfile. so you could explain it better with per-file and per-section or sth like this. maybe we just need to introduce new nomenclaturere